File: iteration2/src/CommandLineInterface/CLIStudent.java
package iteration2.src.CommandLineInterface;

import java.util.*;

import iteration2.src.utils.Util;
import iteration2.src.controllers.StudentController;
import iteration2.src.enums.CourseStatus;
import iteration2.src.interfaces.Color;
import iteration2.src.models.Course;
import iteration2.src.models.SelectedCourse;
import iteration2.src.models.CourseSection;

// TODO: change println to printf so it looks better

public class CLIStudent {

    private StudentController studentController;
    private Scanner scanner;
    private boolean shouldQuit;

    public CLIStudent(StudentController studentController) {
        this.studentController = studentController;
        scanner = new Scanner(System.in);
    }

    public void menuPage() {
        shouldQuit = true;
        while (shouldQuit) {
            Util.clearScreen();

            if (studentController.getNotification() != null) {
                Util.paintTextln(" Notification\n " +
                        "**************", Color.YELLOW);
                for (String string : studentController.getNotification())
                    Util.paintTextln(string, Color.RED);
                studentController.clearNotifications();
            }

            System.out.println("\n\n");
            System.out.println(
                    " Menu\n" +
                            "********\n" +
                            "  1. My Courses\n" +
                            "  2. Log out\n\n" +
                            "Press q to quit");

            String str = scanner.nextLine();

            if (str.equals("1")) {
                showMyCoursesPage();
            } else if (str.equals("2")) {
                break;
            } else if (str.equals("q")) {
                shouldQuit = false;
            } else {
                Util.sendFeedback("Invalid Input: " + str, Color.RED);
            }

        }

    }

    private void showMyCoursesPage() {
        shouldQuit = true;

        while (shouldQuit) {
            Util.clearScreen();
            System.out.println(
                    " My Courses\n" +
                            "**************\n");
            System.out.printf("    %10s    %50s    %15s    Status\n", "Code", "Name", "Section");
            // " Code Name Section Status\n" +
            // " ____ ____ _______ ______");
            System.out.printf("    %10s    %50s    %15s    ______\n", "____", "____", "_______");

            listSelectedCourses();

            System.out.println(
                    "\n\n1. Add Course\n" +
                            "2. Delete Course\n" +
                            "3. Show Timetable\n" +
                            "4. Send To Approval");

            System.out.println("Press b to go back");
            System.out.println("Press q to quit");

            String str = scanner.nextLine();

            try {
                if (str.equals("1")) {
                    showAddCoursePage();
                } else if (str.equals("2")) {
                    System.out.println("Enter the row number of the course you want to delete : ");
                    str = scanner.nextLine();
                    if (Util.validateNumber(str, studentController.getSelectedCourses().toArray())) {
                        if (deleteCourse(str)) {
                            Util.sendFeedback("Course successfully deleted", Color.GREEN);
                        } else {
                            Util.sendFeedback("Course deletion failed", Color.RED);
                        }
                    } else {
                        throw new Exception("Invalid input: " + str);
                    }
                } else if (str.equals("3")) {
                    showTimetablePage();
                } else if (str.equals("4")) {
                    studentController.sendSelectedCoursesToApproval();
                } else if (str.equals("b")) {
                    break;
                } else if (str.equals("q")) {
                    shouldQuit = false;
                } else {
                    throw new Exception("Invalid input: " + str);

                }
            } catch (Exception e) {
                Util.sendFeedback(e.getLocalizedMessage(), Color.RED);
            }

        }
    }

    private void showAddCoursePage() {
        shouldQuit = true;

        while (shouldQuit) {
            Util.clearScreen();
            System.out.println(
                    " Avaliable Courses(To Add)\n" +
                            "**************\n");
            System.out.printf("    %10s    %50s    %15s    %20s    %s\n", "Code", "Name", "Section", "Instructor",
                    "Credit");
            System.out.printf("    %10s    %50s    %15s    %20s    %s\n", "____", "____", "_______", "__________",
                    "______");

            listAvaliableCourseSections();

            System.out.println("\n\npress b to go back");
            System.out.println("press q to quit");
            System.out.println("Enter the row number of the course you want to add : ");

            String str = scanner.nextLine();

            try {
                if (str.equals("b")) {
                    break;
                } else if (str.equals("q")) {
                    shouldQuit = false;
                } else if (Util.validateNumber(str, studentController.getAvaliableCourseSections().toArray())) {
                    if (addCourse(str)) {
                        Util.sendFeedback("Course successfully added", Color.GREEN);
                    } else {
                        Util.sendFeedback("Course addition failed", Color.RED);
                    }
                }

                else {
                    throw new Exception("Invalid input: " + str);
                }

            } catch (Exception e) {
                Util.sendFeedback(e.getLocalizedMessage(), Color.RED);
            }

        }

    }

    private void showTimetablePage() {
        shouldQuit = true;
        while (shouldQuit) {
            Util.clearScreen();

            System.out.println(" Timetable\n" +
                    "***********\n" +
                    " Hours/Days    Monday    Tuesday    Wednesday    Thursday    Friday\n" +
                    " __________    ______    _______    _________    ________    ______\n");
            Util.printTimeTable(studentController.getTimeTable());

            System.out.println("\n\npress b to go back");
            System.out.println("press q to quit");
            String str = scanner.nextLine();

            try {
                if (str.equals("b")) {
                    break;
                } else if (str.equals("q")) {
                    shouldQuit = false;
                } else {
                    throw new Exception("Invalid input: " + str);
                }

            } catch (Exception e) {
                Util.sendFeedback(e.getLocalizedMessage(), Color.RED);
            }
        }
    }

    private boolean addCourse(String str) {
        int rowNumber = Integer.parseInt(Util.getRowNumberFromInput(str));
        CourseSection selectedCourseSection = studentController.getAvaliableCourseSections().get(rowNumber - 1);
        if (studentController.addSelectedCourse(
                new SelectedCourse(selectedCourseSection.findCourseOfCourseSection(), selectedCourseSection))) {
            return true;
        }
        return false;
    }

    private boolean deleteCourse(String str) {
        int rowNumber = Integer.parseInt(Util.getRowNumberFromInput(str));
        if (studentController.removeSelectedCourse(studentController.getSelectedCourses().get(rowNumber - 1))) {
            return true;
        }
        return false;
    }

    private void listAvaliableCourseSections() {
        List<CourseSection> avaliableCourseSections = studentController.getAvailableCourseSections();
        int rowCount = 1;
        Course course;
        for (CourseSection courseSection : avaliableCourseSections) {
            course = courseSection.findCourseOfCourseSection();
            System.out.printf("%d.  %10s    %50s    %15s    %20s    %d\n", rowCount, course.getCourseCode(),
                    course.getCourseName(),
                    courseSection.getSectionCode(), courseSection.getLecturerName(), course.getCourseCredit());
            rowCount++;
        }

    }

    private void listSelectedCourses() {
        int rowCount = 1;
        for (SelectedCourse selectedCourse : studentController.getSelectedCourses()) {

            System.out.printf("%d.  %10s    %50s    %15s    ", rowCount, selectedCourse.getCourse().getCourseCode(),
                    selectedCourse.getCourse().getCourseName(),
                    selectedCourse.getCourseSection().getSectionCode());

            if (selectedCourse.getStatus() == CourseStatus.APPROVED)
                Util.paintText(selectedCourse.getStatus() + "\n", Color.GREEN);
            else if (selectedCourse.getStatus() == CourseStatus.PENDING)
                Util.paintText(selectedCourse.getStatus() + "\n", Color.YELLOW);
            else if (selectedCourse.getStatus() == CourseStatus.DENIED)
                Util.paintText(selectedCourse.getStatus() + "\n", Color.RED);
            else
                System.out.print(selectedCourse.getStatus() + "\n");
            /*
             * System.out.println(rowCount + ". " +
             * selectedCourse.getCourse().getCourseCode() + "\t"
             * + selectedCourse.getCourse().getCourseName() + "\t"
             * + selectedCourse.getCourseSection().getSectionCode() + "\t" +
             * selectedCourse.getStatus());
             */
            rowCount++;
        }
    }

}

File: iteration3/src/CommandLineInterface/CLIStudent.py
from iteration3.src.enums.ApprovalStatus import ApprovalStatus
from iteration3.src.utils.DatabaseManager import DatabaseManager
from iteration3.src.utils.Util import Util
from iteration3.src.interfaces.Color import Color
from iteration3.src.enums.CourseStatus import CourseStatus
from iteration3.src.models.SelectedCourse import SelectedCourse

class CLIStudent(object):
    
    

    def __init__(self, studentController):
        self._studentController = studentController
        self._shouldQuit = True


    def menuPage(self):
        
        while(self._shouldQuit):
            Util.clearScreen()

            if self._studentController.getNotifications() is not None:
                Util.paintTextln(" Notifications\n " +
                                "**************", Color.YELLOW)
                for string in self._studentController.getNotifications():
                    Util.paintTextln(string, Color.RED)
                self._studentController.clearNotifications()


            print("\n\n")
            print(" Menu\n" +
                            "********\n" +
                            "  1. My Courses\n" +
                            "  2. Log out")
            _choice = input()
            if _choice == "1":
                self._showMyCoursesPage()
            elif _choice == "2":
                self._studentController.logout()
                self._shouldQuit = False
            else:
                Util.sendFeedback("Invalid input" + _choice, Color.RED)


    def _showMyCoursesPage(self):
        while(self._shouldQuit):
            Util.clearScreen()
            print("\n\n")
            print(" My Courses\n" +
                "**************\n")
            print("    %-10s    %-50s    %-15s    Status" % ("Code", "Name", "Section"))
            print("    %-10s    %-50s    %-15s    ______\n" % ("____", "____", "_______"))


            self._listSelectedCourses()

            print( "\n\n1. Add Course\n" +
                    "2. Delete Course\n" +
                    "3. Show Timetable\n" +
                    "4. Send To Approval\n")
            
            print("Press b to go back")
            print("Press q to quit\n")

            _choice: str = input()

            try:
                if _choice == "1":
                    if DatabaseManager.getInstance().getConstraints().get(4) == "False"  and  DatabaseManager.getInstance().getConstraints().get(2) == "False":
                        Util.sendFeedback("You are not in registration nor add-drop week. You cannot select a new course.", Color.RED)
                    elif self._studentController.getApprovalStatus() == ApprovalStatus.FINALIZED_REGISTRATION:
                        Util.sendFeedback("Your registration has been finalized. You cannot select a new course.", Color.RED)
                    else:
                        self._showAddCoursePage()

                elif _choice == "2":
                    if DatabaseManager.getInstance().getConstraints().get(4) == "False"  and  DatabaseManager.getInstance().getConstraints().get(2) == "False":
                        Util.sendFeedback("You are not in registration nor add-drop week. You cannot delete a course.", Color.RED)
                    elif self._studentController.getApprovalStatus() == ApprovalStatus.FINALIZED_REGISTRATION:
                        Util.sendFeedback("Your registration has been finalized. You cannot delete a course.", Color.RED)                        
                    else:
                        _choice = input("Enter the row number of the course you want to delete: ")
                        if(Util.validateNumber(_choice, self._studentController.getSelectedCourses()) ): 
                            self._deleteCourse(_choice)
                            Util.sendFeedback("Course deleted successfully", Color.GREEN)
                        else:
                            raise Exception("Invalid input " + _choice)
                        
                elif _choice == "3":
                    self._showTimeTablePage()

                elif _choice == "4":
                    self._studentController.sendSelectedCoursesToApproval()
                    Util.sendFeedback("Selected courses sent to approval successfully", Color.GREEN)
                elif _choice == "b":
                    break

                elif _choice == "q":
                    self._studentController.logout()
                    self._shouldQuit = False
                    
                else:
                    raise Exception("Invalid input " + _choice)
                    
            except Exception as e:
                Util.sendFeedback(str(e), Color.RED)
    
    def _showAddCoursePage(self):

        self._shouldQuit = True
        while(self._shouldQuit):
            Util.clearScreen()
            print("\n\n")
            print(" Add Course\n" +
                "**************\n")
            print("    %-10s    %-50s    %-15s    %-20s    %s\n" % ("Code", "Name", "Section", "Instructor", "Credit"))
            print("    %-10s    %-50s    %-15s    %-20s    %s\n" % ("____", "____", "_______", "__________", "______"))


            self._listAvailableCourseSections()
            
            print("\nPress b to go back")
            print("Press q to quit\n")
            

            _choice = input("Enter the row number of the course you want to add: ")

            try:
                if _choice == "b":
                    break

                elif _choice == "q":
                    self._studentController.logout()
                    self._shouldQuit = False

                else:
                    if (Util.validateNumber(_choice, self._studentController.getAvailableCourseSections())):
                        self._addCourse(_choice)
                        Util.sendFeedback("Course added successfully", Color.GREEN)
                    else:
                        raise Exception("Invalid input: " + _choice)
            except Exception as e:
                Util.sendFeedback(str(e), Color.RED)
            
    def _showTimeTablePage(self):
        self._shouldQuit = True
        while(self._shouldQuit):
            Util.clearScreen()
            Util.printTimeTable(self._studentController.getTimeTable())
            print("\n\nPress b to go back")
            print("Press q to quit\n")

            _choice = input()

            try:
                if _choice == "b":
                    break
                elif _choice == "q":
                    self._studentController.logout()
                    self._shouldQuit = False
                else:
                    raise Exception("Invalid input" + _choice)
            except Exception as e:
                Util.sendFeedback(str(e), Color.RED)
    
    def _addCourse(self, str):
        rowNumber = int(str) 
        selectedCourseSection = self._studentController.getAvailableCourseSections()[(rowNumber-1)]
        if self._studentController.addSelectedCourse( SelectedCourse( selectedCourseSection.findCourseOfCourseSection(), CourseStatus.DRAFT, selectedCourseSection ) ):
            return True

        return False
    
    def _deleteCourse(self, str):
        rowNumber = int(str)
        if self._studentController.removeSelectedCourse( self._studentController.getSelectedCourses()[rowNumber-1] ):
            return True
        
        return False
    
    def _listAvailableCourseSections(self):
        rowCount = 1
        for courseSection in self._studentController.getAvailableCourseSections():
            course = courseSection.findCourseOfCourseSection()
            print(f"{rowCount}.  {course.getCourseCode():<10}    {course.getCourseName():<50}    {courseSection.getSectionCode():<15}    {courseSection.getLecturerName():<20}    {course.getCourseCredit()}")
            rowCount += 1

    def _listSelectedCourses(self):
        rowCount = 1
        for selectedCourse in self._studentController.getSelectedCourses():
            print(f"{rowCount}.  {selectedCourse.getCourse().getCourseCode():<10}    {selectedCourse.getCourse().getCourseName():<50}    {selectedCourse.getCourseSection().getSectionCode():<15}    ",  end="")            
            if selectedCourse.getStatus() == CourseStatus.APPROVED:
                Util.paintTextln(f"{selectedCourse.getStatus().value}", Color.GREEN)
            elif selectedCourse.getStatus() == CourseStatus.PENDING:
                Util.paintTextln(f"{selectedCourse.getStatus().value}", Color.YELLOW)
            elif selectedCourse.getStatus() == CourseStatus.DENIED:
                Util.paintTextln(f"{selectedCourse.getStatus().value}", Color.RED)
            elif selectedCourse.getStatus() == CourseStatus.ACTIVE:
                Util.paintTextln(f"{selectedCourse.getStatus().value}", Color.GREEN)
            elif selectedCourse.getStatus() == CourseStatus.APPROVED_FINALIZED:
                Util.paintTextln("APPROVED FINALIZED", Color.GREEN)
            elif selectedCourse.getStatus() == CourseStatus.DENIED_FINALIZED:
                Util.paintTextln("DENIED FINALIZED", Color.RED)
            else:
                Util.paintText(f"{selectedCourse.getStatus().value}\n", Color.DEFAULT)
            
            rowCount += 1

    